### Data Frames

This is the main reason we use `python` for data analysis. The `data frame` data structure is incredibly useful for working with data. Just like `arrays` were not part of base `python`, however, neither is the `data frame`.

First let's import the packages we need. Our new `data structure` will come via the `pandas` package. 

In [1]:
import numpy as np
import pandas as pd

Now, let's import some data. There's a folder titled `data` in Canvas Files section. In that folder is a file called `us_state_pop_2024.csv`. Download that file and move it to the same folder that your jupyter notebook is in.

Now, read that file into python using the following command:

In [2]:
df=pd.read_csv('us_state_pop_2024.csv')

In [4]:
len(df)

51

In [8]:
#df.head()

If you're trying to keep your folders and files organized on your computer, you might want to have a separate folder for data files. For example, maybe you have a folder called `bsds100` where your jupyter notebooks are. Then inside that folder you can create another folder called `data` and put our file inside. Then we can read the file in as follows:

In [19]:
df=pd.read_csv('data/us_state_pop_2024.csv')

If you get an error, the most likely cause is that python can't find the file, so make sure it lives in the same folder with your notebooks. 

Note 1: The file we read in is a `.csv` file which stands for *comma separated values*, and is a common format for data files.

Note 2: The `pd.read_csv()` command is a function that reads the file in as a data frame.  We'll talk about it more in a minute. Note, however, that we also assigned the file (via the `=`) to a variable named `df` (for data frame) but we certainly could have used a different name.  

Let's look at the file. There's a useful method for data frames (data frames are, of course, objects) that let's us peek at part of the file:

In [7]:
df.head()

State  Population  Adult_Population  sq_miles Coast
0     Alabama     5157699           4022842     52420    GM
1      Alaska      740133            565186    665384    PO
2     Arizona     7582384           5994209    113990   NaN
3    Arkansas     3088354           2386510     53179   NaN
4  California    39431263          31012711    163695    PO

Data frames are used for *tabular* data: Like an Excel spreadsheet or a Google sheet.

Tabular data consists of *ROWS* and *COLUMNS*. (So in some ways it's like a two dimensional numpy array, BUT, much more flexible and general as we'll see.)  

In [11]:
df.shape

(51, 5)

Our data frame has 51 rows and 5 columns.

Notice that each row and column has a *name*. The row names are in bold on the left side of the dataframe. In this case, they seem to be the integers 0,1,2,3,4,.... The names for the rows in a data frame is called the `index` of the data frame. Even though it's the case here, in general the index doesn't have to be integers so that's different from numpy arrays.

The column names are in bold along the top row. They are words (or strings) that seem to be describing the information stored in each column.

In [12]:
df.index

RangeIndex(start=0, stop=51, step=1)

In [13]:
df.columns

Index(['State', 'Population', 'Adult_Population', 'sq_miles', 'Coast'], dtype='object')

Note that pandas thinks of the columns as an `index` too. That's becuase the names of the columns are like labels for the columns just like the numbers are the labels for each row.

Indeed, the `index` of `df` runs from 0 to 51 in steps of 1 and the four column names are listed.

If we want to see more of the data frame we can ask (though at some point if there's too much output it will get hidden):

In [14]:
df.head(10)

State  Population  Adult_Population  sq_miles   Coast
0      Alabama     5157699           4022842     52420      GM
1       Alaska      740133            565186    665384      PO
2      Arizona     7582384           5994209    113990     NaN
3     Arkansas     3088354           2386510     53179     NaN
4   California    39431263          31012711    163695      PO
5     Colorado     5957493           4744328    104094     NaN
6  Connecticut     3675069           2947242      5543      AO
7     Delaware     1051917            838204      2489      AO
8           DC      702250            572419        68     NaN
9      Florida    23372215          18872523     65758  AO, GM

(Also note that in row 8 we have the District of Columbia so the size of our dataframe makes sense.)

In [15]:
df.tail()

State  Population  Adult_Population  sq_miles Coast
46       Virginia     8811195           6927764     42775    AO
47     Washington     7958180           6303143     71298    PO
48  West Virginia     1769979           1421615     24230   NaN
49      Wisconsin     5960975           4719976     65496    GL
50        Wyoming      587618            459626     97813   NaN

(It seems that our data is in alphabetical order.)

Data Frames consist of **observations** (rows) of **variables** (columns). (Not exacly the same as variables in python -- names of objects -- usually no confusion based on context.)

Each row is a state which we can *measure* from the perspective of the state's population, it's adult population (the number of residents over the age of 18), and it's geographic size (measured in square miles). 

Note that as we move between among the rows, we see different states, and the value of the particular variable (like **Population** or **sq_miles**) will change (vary over the observations).

Another data set might have individual people as the observations, and their name, height, and hair color as variables.

Variables can be *quantitative* (Population or height) or *categorical* (Coast or hair color)

(What is the average population or size (in square miles) makes sense. What is the average Coast value of hair color does not make sense.)

For now, let's throw away the `Coast` column. We can come back to it later on.

In [16]:
del df['Coast']

In [17]:
df.head()

State  Population  Adult_Population  sq_miles
0     Alabama     5157699           4022842     52420
1      Alaska      740133            565186    665384
2     Arizona     7582384           5994209    113990
3    Arkansas     3088354           2386510     53179
4  California    39431263          31012711    163695

Some questions we might want to answwer...

1. Can we change the units of the population columns so they are given in millions of people.
1. What states are the most populous?
2. What states have the largest percentage of children?
3. What states are geographically largest?
4. What states are most densely populated?
   

We will return to these questions in a sec. 

#### Series

A series is a new data structure from pands that worth discussing for a minute. You can (basically) think of a Series as a one dimensional `numpy array` but, with an index that does not need to be an integer.

Example: 

In [18]:
ser1=pd.Series([1,1.5,2,7])

In [19]:
ser2=pd.Series([1,1.5,2,7],index=['Charlie','Bob','Amy','Sally'])

In [20]:
ser1

0    1.0
1    1.5
2    2.0
3    7.0
dtype: float64

In [21]:
ser2

Charlie    1.0
Bob        1.5
Amy        2.0
Sally      7.0
dtype: float64

We can use the index labels to access the series values.

In [22]:
ser1[1]

np.float64(1.5)

In [23]:
ser1[1:]

1    1.5
2    2.0
3    7.0
dtype: float64

In [24]:
ser1/10

0    0.10
1    0.15
2    0.20
3    0.70
dtype: float64

In [25]:
ser2['Bob']

np.float64(1.5)

In [26]:
ser2['Bob':]

Bob      1.5
Amy      2.0
Sally    7.0
dtype: float64

You can think of each column in a data frame as a series, and you can think of a data frame as a collection of series that are 'aligned' along the index and glued together. That means that the column values are referring to the same observation (row/index).

In [27]:
type(df['Population'])

pandas.core.series.Series

In [28]:
type(df['sq_miles'])

pandas.core.series.Series

In [29]:
df['Population'][2]

np.int64(7582384)

In [30]:
df['sq_miles'][2]

np.int64(113990)

In [31]:
df['State'][2]

'Arizona'

In [32]:
df.loc[2]

State               Arizona
Population          7582384
Adult_Population    5994209
sq_miles             113990
Name: 2, dtype: object

Note: We accessed the column of the data frame using `df['column_name']`

We accessed a row of the data frame using the .loc method: `df.loc['row_name']`. Here the rows were named 0,1,2,...

In `ser2` (a Series, like a one-column data frame) the rows (which only have one value in them) were names: 

In [35]:
ser2.head()

Charlie    1.0
Bob        1.5
Amy        2.0
Sally      7.0
dtype: float64

In [34]:
ser2.loc['Bob']

np.float64(1.5)

If we wanted to access a `cell` (like the entry in the row labeled 3 of dataframe df, and the column named 'Population':  

In [36]:
df.loc[3,'Population'] #loc = location

np.int64(3088354)

In [48]:
df.loc[2]

State               Arizona
Population          7582384
Adult_Population    5994209
sq_miles             113990
Name: 2, dtype: object

If `ser2` is the age of the kids named in the index, and we had another series that gives the number of siblings, like:

In [41]:
sibs=pd.Series([2,2,3,0],index=['Charlie','Bob','Amy','Sally'])

In [42]:
sibs.head()

Charlie    2
Bob        2
Amy        3
Sally      0
dtype: int64

Then we could make a dataframe whose columns are `age` and `siblings` and whose index is the kids' names, which is the common (aligned) index for both series. 

In [43]:
df_with_series = pd.DataFrame(
    { 'Age':ser2,
      'Siblings':sibs
    })

In [47]:
df_with_series

Age  Siblings
Charlie  1.0         2
Bob      1.5         2
Amy      2.0         3
Sally    7.0         0

In [49]:
df_with_series.loc['Bob']

Age         1.5
Siblings    2.0
Name: Bob, dtype: float64

In [50]:
df_with_series['Age']

Charlie    1.0
Bob        1.5
Amy        2.0
Sally      7.0
Name: Age, dtype: float64

What happens if you try to make a dataframe where the indices aren't aligned?

In [51]:
snack_ser=pd.Series(['Carrots','Chips','Nuts','Yogurt','Bread'],index=['Charlie','Bob','Amy','Sally','Ryan'])

In [52]:
test=pd.DataFrame(
    { 'Age':ser2,
      'Siblings':sibs,
      'Snack':snack_ser
    })

In [53]:
test

Age  Siblings    Snack
Amy      2.0       3.0     Nuts
Bob      1.5       2.0    Chips
Charlie  1.0       2.0  Carrots
Ryan     NaN       NaN    Bread
Sally    7.0       0.0   Yogurt

One more quick example, here's a dataframe:

In [67]:
sc=pd.read_csv('scrabble.csv')

That was the right code for me but maybe not for you. Read in the `scrabble.csv` data file.

Let's take a look:

In [68]:
sc.head()

Letter  Value
0      a      1
1      b      3
2      c      3
3      d      2
4      e      1

In [59]:
type(sc)

pandas.core.frame.DataFrame

What kind of object is this?

Now run the code below. What did it do? Which is better, `sc` or `sc_ser`? Why? 

In [63]:
sc_ser=pd.Series(sc['Value'].values,index=sc['Letter'])

In [70]:
sc_ser['r']

np.int64(1)

So, recapping:

A data frame is a collection of columns (variables, series) glued together so that they align along rows (observations, index).

We typically try to think of a data frame in terms of its columns, or sometimes, in terms of it's rows (stacked together horizontally). That way we're thinking about variables or observations.

We can access only some of the columns by passing the column names in a list to the dataframe.

Here's an example.

In [71]:
df[['Population','sq_miles']].head()

Population  sq_miles
0     5157699     52420
1      740133    665384
2     7582384    113990
3     3088354     53179
4    39431263    163695

Since this subset of our original frame has only numerical columns, we can apply methods like .mean() or .max().

In [72]:
df[['Population','sq_miles']].mean()

Population    6.668843e+06
sq_miles      7.444596e+04
dtype: float64

These are in scientific notation. The 6.668843e+06 means 6.668843 times 10 to the 6th power.

In [73]:
df[['Population','sq_miles']].max()

Population    39431263
sq_miles        665384
dtype: int64

What happens if you try these methods (.mean() and .max()) on the whole df dataframe?

In [75]:
#df.mean()

It doesn't work because there are different types of data and it wouldn't make sense to take the mean or median of a State name for example.

### A word on filtering

Sometimes you want to pick out certain elements of an array (or rows of a data frame).

In [76]:
arr=np.array([11.2,0.34,2.54,0.98,-1.37, 9.21])

In [77]:
arr

array([11.2 ,  0.34,  2.54,  0.98, -1.37,  9.21])

We know we can access elements of the array py passing the index in brackets.

In [78]:
arr[3]

np.float64(0.98)

But we can also pass a list of the indices we want:

In [79]:
arr[[0,3,4]]

array([11.2 ,  0.98, -1.37])

Or a boolean list indicating the indices we want (and those we don't).

In [80]:
arr[[True,False,False,True,True,False]]

array([11.2 ,  0.98, -1.37])

Suppose I want the entries in `arr` that have values greater than 1.

A convenient way to do it is to create a boolean list of the indices where elements are greater than 1, and then pass that list to the array.

In [81]:
arr>1

array([ True, False,  True, False, False,  True])

In [82]:
arr[arr>1]

array([11.2 ,  2.54,  9.21])

Since data frames are built on arrays, the same strategy will work for a data frame. For example, if we want all the states with more than 10 million people:

In [83]:
df[df['Population']>10000000]

State  Population  Adult_Population  sq_miles
4       California    39431263          31012711    163695
9          Florida    23372215          18872523     65758
10         Georgia    11180878           8640127     59425
13        Illinois    12710158          10012697     57914
22        Michigan    10140459           8031116     96714
32        New York    19867248          15884969     54555
33  North Carolina    11046024           8685722     53819
35            Ohio    11883304           9308934     44826
38    Pennsylvania    13078751          10448930     46054
43           Texas    31290831          23625608    268596

Note that `df['Population']>10000000` gives the index values (rows) we want as a boolean list which we then passed to the data frame.

In [86]:
example = df['Population']>10000000
example.head()

0    False
1    False
2    False
3    False
4     True
Name: Population, dtype: bool

OK. Let's answer some of the questions we asked earlier about our data frame of state populations.

Question 5. What states are the most densely populated? How should we define density?

Now we will reap the benefits of the flexibility inherent in using numpy arrays to do arithmetic, and the fact that pandas objects like series are built on top of arrays. (Arrays are under the hood, so array style operations should work.)

Quick example/reminder

In [87]:
arr1=np.array([1,2,3])
arr2=np.array([4,5,6])

In [88]:
result_array=arr1/arr2 

What should this be?

In [89]:
result_array

array([0.25, 0.4 , 0.5 ])

Gameplan for Question 5. 

For each state (row), 

divide it's population (the value of the 'Population' column in that row) by it's size (the value of sq_miles' in that row) 

and 

keep track of that value -- 

maybe even make the result (pop/size) the row value of a new column called 'density'.

In [101]:
density = df['Population']/df['sq_miles']

In [102]:
type(density)

pandas.core.series.Series

In [103]:
df["Density"] = density

In [105]:
df.head()

State  Population  Adult_Population  sq_miles     Density
0     Alabama     5157699           4022842     52420   98.391816
1      Alaska      740133            565186    665384    1.112340
2     Arizona     7582384           5994209    113990   66.517975
3    Arkansas     3088354           2386510     53179   58.074691
4  California    39431263          31012711    163695  240.882513

In [109]:
df_sorted = df.sort_values("Density", ascending = False)

In [111]:
df_sorted.head(1)

State  Population  Adult_Population  sq_miles       Density
8    DC      702250            572419        68  10327.205882

Finally, we'd probably want to sort the values by the entries in that column to find the biggest or smallest. (In this case we might not need to but imagine if you had a dataframe with thousands (or more) of rows.

One last comment. If you want to save a dataframe you've created (maybe the states dataframe with the density column) you can write itr to your computer as follows:

In [112]:
df.to_csv('density_of_USA_states.csv')

It should go to the same folder within which your jupyter notebook is running.

## Playground

1. Finish Implementing the gameplan for question 5 above.
2. Answer questions 1-4 about the states data frame we called `df`.
3. Import the titanic dataset `titanic.csv`.
4. How many people survived? How many were men? Women?
5. What was the average age of survivors? Non survivors?
6. What was the survival rate? 
7. Generate some questions you might be curious about regarding the titanic. 
8. Which questions can you answer? Are there any techniques you need but don't have yet?

In [148]:
#1:
df_sorted.head(1)

State  Population  Adult_Population  sq_miles       Density
8    DC      702250            572419        68  10327.205882

The state with the highest density of population is DC.

In [138]:
#3.
titanic_df = pd.read_csv("titanic.csv")
titanic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [145]:
#4.
survived = sum(titanic_df['Survived'])
print(f'A total of {survived} people survived.')


males = titanic_df[titanic_df['Sex'] == 'male']
males_survived = sum(males['Survived'])
print(f'A total of {males_survived} males survived.')

females = titanic_df[titanic_df['Sex'] == 'female']
females_survived = sum(females['Survived'])
print(f'A total of {females_survived} females survived.')

A total of 342 people survived.
A total of 109 males survived.
A total of 233 females survived.


In [ ]:
titanic_df